### ISSUE TITLE
[Feature] Chronos-2 fine-tuning support

### ISSUE THREAD

**Issue by daidahao** [unit8co/darts#2964]
**Is your feature request related to a current problem? Please describe.**
It might be a good time to think about fine-tuning support of Chronos-2 since it has been supported in [Amazon's repo](https://github.com/amazon-science/chronos-forecasting/) and [autogluon](https://github.com/abdulfatir/autogluon/blob/master/timeseries/src/autogluon/timeseries/models/chronos/chronos2.py).

In Amazon's repo, the default fine-tuning strategy is `"full"` so all model weights will be updated. But in autogluon (Amazon's time series library), the default fine-tuning strategy is [LoRA](https://huggingface.co/docs/peft/en/package_reference/lora) which means only parts of the model will be updated and in fact some low-rank matrices would injected into the network as weights. PEFT is required for LoRA but not for `"full"`.

There might be a challenge of implementing lora in Darts since 1) it would add `peft` as a dependency and 2) it is unclear whether peft is compatible with Lightning.

**Describe proposed solution**
Enable the `_allows_finetuning` in `Chronos2Model` for initial testing. However, there might be a problem with the normalized training loss as noted in the comments:

```
    # TODO: fine-tuning support w/ normalized loss
    # Currently, Darts own `RINorm` is not used as Chronos-2 has its own implementation. Major differences
    # 1. Chronos-2 `RINorm` normalizes both target and covariates, while Darts normalizes target only.
    # 2. Chronos-2 `RINorm` additionally applies `arcsinh` transformation after standardization
    # 3. Chronos-2 uses normalized values for loss computation, while Darts uses denormalized values.
    # We need to think about how best to implement Chronos-2 `RINorm` in `io_processor()` without
    # breaking existing behavior, while also allowing fine-tuning with normalized loss.
```

**Describe potential alternatives**
A clear and concise description of any alternative solutions or existing features that might solve it.

**Additional context**

**Comment by dennisbader** [comment]
Hi @daidahao and thanks for adding this feature request. @Kurokabe will do an investigation of how we could enable support for fine-tuning.

**Comment by Kurokabe** [comment]
Hi @daidahao and @dennisbader 👋
I wanted to share an initial proposal for enabling fine-tuning support in Darts / Chronos and gather feedback.

The main work-in-progress is pushed on the [finetuning  branch](https://github.com/unit8co/darts/compare/master...finetuning)

### High-level idea

The core change is a refactoring of the internal Chronos model:

* The concrete Chronos implementation has been extracted into its own module
* The PyTorch Lightning module (`_Chronos2PLModule`) is now clearly separated from the model implementation itself

This separation keeps the training logic (PL module) independent from the model definition, and makes it straightforward to wrap the Chronos model with PEFT / LoRA-style adapters.

With this structure, fine-tuning becomes a relatively thin wrapper around the concrete Chronos model rather than being tightly coupled to Lightning internals.

### PEFT / dependency discussion

The most natural solution for LoRA-based fine-tuning here would be **🤗 PEFT**:

* Actively maintained by the Hugging Face team
* Clean and well-documented API
* Supports:

  * Adapter-only saving & loading (only LoRA matrices, not full model weights)
  * Easy switching between base and adapted models
  * Multiple adapter configurations

These features are especially useful for:

* Reducing saved model size (only the adapter needs to be saved)
* Faster load times (while the core model still needs to be loaded once, switching between use-case-specific adapters is much faster than reloading full model weights)
* Sharing fine-tuned adapters without redistributing full models

A key concern is the dependency footprint:

* `peft` would become a new dependency
* `peft` requires `transformers`, which is not exactly lightweight

One possible mitigation is to make PEFT an **optional dependency**, e.g.:

```bash
pip install darts[peft]
```

This would allow users who are interested in fine-tuning to opt in, while keeping the default installation lightweight for inference-only or forecasting use cases.

While there *are* lighter-weight LoRA alternatives (e.g. [https://github.com/changjonathanc/minLoRA](https://github.com/changjonathanc/minLoRA)), I currently lean toward PEFT due to:

* Long-term maintenance and stability
* Better ergonomics
* Built-in adapter lifecycle management (save/load/merge)
* Wider adoption and ecosystem support

That said, this is very much open for discussion.

### Regular fine-tuning (without LoRA)

Beyond LoRA-based approaches, **standard fine-tuning** would also be possible with this refactoring:

* Users could manually freeze/unfreeze subsets of the Chronos model parameters (e.g. freeze the backbone and fine-tune only specific layers)
* Alternatively, we could provide a small convenience interface for this (e.g. layer-wise freezing helpers), to make regular fine-tuning easier and less error-prone

This keeps the door open for fine-tuning workflows that don’t rely on PEFT at all.

### Loss / training details

I haven’t yet dug deeply into the **normalized loss** details. One possible approach could be to add a **custom loss hook** directly in `_Chronos2PLModule`, but this would need a closer look and discussion to ensure consistency with existing training behavior.

---

This is meant as a **starting point** rather than a finalized design. I’d love feedback on:

* The refactoring approach
* Whether PEFT as an *optional* dependency makes sense
* Expectations around regular vs LoRA-based fine-tuning

Happy to clarify or iterate on any part of this proposal 🙂

**Comment by daidahao** [comment]
@Kurokabe 
Thank you very much for the efforts! It would be great to see Chronos-2 fine-tuning being supported in Darts and the potential it would unlock.

That said, I am a bit skeptical of refactoring `_Chronos2Module` and creating `FoundationPLModule` just to support PEFT (LoRA being one of the methods) for foundation models. While I see the potential benefits of supporting PEFT, I also see three major concerns with this approach:
1. **"Optional" Dependencies**: Using any `FoundationModel` (Chronos-2 and future models) now requires installing `peft` and `transformers`. As you have said, the latter is not quite lightweight. It is also subject to future API changes. Previously, `FoundationModel` can be used out-of-the-box with pip `darts` and its main dependencies (`huggingface-hub`, `safetensors`) are lightweight except torch itself. The new changes would make `peft` and `transformers` effectively compulsory for foundation models.
2. **PEFT Configuration**: Adding `peft_config` argument to `FoundationModel` means that all future models would support peft-style fine-tuning regardless of their model size or memory footprint. Since users might not be familiar with LoRA, it means that we would need to at least supply default `peft_config` for all models for user reference. Chronos-2 provides default `peft_config` we can use directly, but we might not be as confident for new models.
3. **Benefits vs Risks**: The benefits of PEFT over full model fine-tuning are unknown to us as of now. How much memory or training time are we saving with LoRA on a non-trivial dataset? Do the savings justify setting up `peft` (and `transformers`) or user time spent spent on finding a `peft_config`? Is there any accuracy discrepancy between LoRA and full fine-tuning for Chronos-2? Also, if memory or training time is a concern for full fine-tuning of Chronos-2, can we use the smaller checkpoint #2962 instead of LoRA?

Let me know what you think and I am happy to help with any PR or code review.

**Comment by Kurokabe** [comment]
Hi @daidahao and thanks for your insights!

I agree that forcing users to rely on peft is probably not the best approach. Ultimately, the choice of fine-tuning strategy (LoRA vs full fine-tuning) should be left to the end user. What can still be valuable, though, is offering a simpler and more flexible interface for fine-tuning foundation models, without introducing additional mandatory dependencies.

In that regard, I think the idea of introducing a FoundationPLModule to decouple the PLForecastingModule from concrete implementations is still beneficial, even if we remove the hard dependency on peft. At a high level, this could look something like:

```python
# Ensures that FoundationPLModule exposes a `model` attribute,
# which users can access and wrap themselves if needed (e.g. with PEFT)
class FoundationPLModule(PLForecastingModule):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.model: nn.Module

class _Chronos2PLModule(FoundationPLModule):
    def __init__(
        self,
        d_model: int = 512,
        d_kv: int = 64,
        d_ff: int = 2048,
        num_layers: int = 6,
        ...
        **kwargs,
    ):
        ...
    
    def forward(...):
        ...
        return self.model.forward(...)

class Chronos2Model(FoundationModel):
    ...
    def _create_model(self, train_sample: TorchTrainingSample) -> FoundationPLModule:
        pl_module_params = self.pl_module_params or {}
        model = self.hf_connector.load_model(
            module_class=_Chronos2PLModule,
            pl_module_params=pl_module_params,
        )
        return model
```

This way, users can directly access and modify the underlying model if they want to apply PEFT (or anything else), without Darts enforcing a specific fine-tuning method:

```python
model = Chronos2Model(...)

from peft import get_peft_model # This could be installed by the user, and not be a darts dependency

model.init_model(...)  # builds the internal PL module
# model.model.model refers to the underlying nn.Module exposed by the PL module
model.model.model = get_peft_model(model.model.model, lora_config)

model.fit(...)  # fine-tuning now uses LoRA
```

Currently, the model creation is done within the `fit` method. We could imagine a way to manually initialize the model, allowing the user to handle the model internals (in order to wrap it with peft, or freeze some layers) before performing the fine-tuning with `fit`. 

Admittedly, accessing the internal model via `model.model.model` is not ideal, but it allows an easy wrapping of the underlying model with peft (otherwise the PLModule would be modified and `model.fit` wouldn't work anymore). It would be possible to hide this behind a small helper on FoundationModel, e.g.:

```python

class FoundationModel:
   def get_torch_model(self) -> nn.Module:
      return self.model.model
```

Concerning the normalized loss, if I understand correctly, there is a difference between how the loss is computed in the original Chronos implementation and Darts:

In Chronos:
```python
# Inside Chronos2Model.forward:
# Step 1: compute the hidden_state, internally the context and future_covariates are normalized with self.instance_norm
# Step 2: compute loss
# Step 3: denormalize quantile_preds with self.instance_norm.inverse
# Step 4: return loss and the denormalized quantile_preds
```
Whereas in Darts:
```python
# Inside PLForecastingModule._train_val_step
# Step 1: compute the model output, internally the context and future_covariates are normalized with self.instance_norm
#         then denormalized with self.instance_norm.inverse
# Step 2: compute the loss
# Step 3: return loss
```

So we need to find a way for Chronos-2 fine-tuning to compute the loss on normalized values, in order to stay consistent with the original Chronos implementation, while keeping the existing Darts behavior unchanged for other models. I’m not sure yet whether this would require changes to io_processor, especially since it seems to behave differently and is currently not used by Chronos-2.

Looking forward to your input and I wish you a Merry Christmas 🎄 🙂

**Comment by daidahao** [comment]
Hi @Kurokabe 

Thank you for your suggestions here. I really appreciate your efforts of making peft a fine-tuning option for users. After much thought, I do have two concerns for the decoupling:

1. If the user wraps a model with `get_peft_model()`, will there be any compatibility issues between peft and lightning? https://github.com/Lightning-AI/pytorch-lightning/discussions/17878 After all, PEFT is designed to work best within HuggingFace ecosystem, but not necessarily Lightning.
2. How would the user load or save a PEFT model in Darts? Would functions like `model.load()` and `model.save()` still work with a PEFT model? My understanding is that a `peft_config` is required to wrap `nn.Module` first before loading a checkpoint.

Overall, I am still neutral on whether PEFT (or at the very least, LoRA) must be supported for foundation models like Chronos-2. However, if we must implement LoRA as a fine-tuning option, I would suggest we could implement LoRA in a Lightning/PyTorch-native way here: https://lightning.ai/docs/overview/finetune-models/lora-finetuning

For the `io_processor`, you are right, we don't need to change it since `RINorm`  behaves differently than the one in Chronos-2. My comments at the time was just me thinking out loud if merging those two `RINorm` is possible. Other foundation models like TimesFM 2.5 also have (slightly) different `RINorm` so a unified `RINorm` is probably unlikely anyway.

For the choice of normalised loss, I now believe it might not be necessary for Chronos-2 since the outputs are de-normalised and Darts always computes loss on the original scale (see `_compute_loss()` in `PLForecastingModule`). We just need to verify whether such "de-normalised" fine-tuning is numerically stable while still improving accuracy. (It would mean, of course, the fine-tuning loss is different than in chronos repo.) What do you think?

Merry Christmas and happy new year to you too! 🎆

**Comment by Kurokabe** [comment]
Happy new year @daidahao, and apologies for the slow reply.

I have more time now to get back to this issue. I’ll first investigate whether I can reproduce or better understand the issues when combining `peft` and `lightning`, especially around adapter save/load, and see what the main blockers are on our side.

From a library-design perspective, my preference would be to let users rely on their fine-tuning framework of choice (e.g. `peft`) rather than reimplementing LoRA logic directly in Darts. This would avoid us having to maintain custom LoRA layers and related logic, and would also give users more flexibility to plug in other fine-tuning approaches that wrap an existing model. Ideally, `peft` would remain a user-side dependency rather than a hard dependency of Darts, with documented support and known limitations.

If it turns out that there are fundamental or hard-to-work-around issues (like the ones you mentioned) that make this approach unreliable, then we could consider alternative solutions. One possible direction could be something closer to the Lightning callback approach used in frameworks like NeMo (see [https://docs.nvidia.com/nemo-framework/user-guide/24.09/nemo-2.0/features/peft.html](https://docs.nvidia.com/nemo-framework/user-guide/24.09/nemo-2.0/features/peft.html). However, I think fully integrating and maintaining our own LoRA implementation inside Darts should probably be a last resort, as it can easily become error-prone and increase long-term maintenance cost.

I’ll report back once I’ve had a chance to dig deeper into the current limitations and see what’s realistically feasible :)

**Comment by daidahao** [comment]
@Kurokabe Thank you for the update.

The NeMo approach is an interesting use of Lightning callbacks and seems compatible with Darts TFM and PL modules. Probably worth a try should the other approaches have some roadblocks.

**Comment by Kurokabe** [comment]
@daidahao 
I used Lightning callbacks to handle three different fine-tuning strategies:

* **Full weight fine-tuning**
* **Partial fine-tuning with frozen layers**
* **PEFT fine-tuning using LoRA via the `peft` library**

For each of these methods, there is a complete example covering fine-tuning, saving, and loading in the following notebook:
[https://github.com/unit8co/darts/blob/finetuning/examples/26-Chronos-2-finetuning-examples.ipynb](https://github.com/unit8co/darts/blob/finetuning/examples/26-Chronos-2-finetuning-examples.ipynb)

The main design choice was to rely on callbacks to control model behavior (e.g., freezing layers or wrapping the model with a PEFT adapter), while still delegating the LoRA implementation to the `peft` library itself. This avoids re-implementing PEFT/LoRA logic inside Darts and keeps the integration relatively lightweight.

At the moment, these callbacks live in `darts/models/forecasting/foundation_model.py`, but they could definitely be moved elsewhere (e.g. `darts/utils/callbacks.py` or ideally `darts/utils/callbacks/fine_tuning.py` and the initial `callbacks.py` could be moved to `darts/utils/callbacks/progress_bar.py`).

One open question is whether we want to keep `PeftCallback` inside Darts, since it introduces an internal dependency on `peft` (even though it is only imported when used). It can, however, serve as a concrete, working example of how an external library can be integrated to enable fine-tuning, which might be useful for users extending Darts in similar ways.

Regarding your previous point (“We just need to verify whether such ‘de-normalised’ fine-tuning is numerically stable while still improving accuracy…”), in my small-scale tests, even though the fine-tuning loss differs from the official Chronos-2 implementation, fine-tuning still led to improved accuracy. That said, more rigorous testing would definitely be beneficial.

Happy to hear your feedback :)

**Comment by daidahao** [comment]
@Kurokabe 
Thanks for the implementation. I especially appreciate the support for three fine-tuning strategies via Lightning callbacks (nice work!).

That said, I remain neutral as to whether LoRA has to be supported at the expense of added complexity and dependencies. My main worry is that:

1) Not many users are familiar with Lightning callback mechanisms and exposing callback APIs for fine-tuning could be confusing at first look.

2) The example notebook shows that the fine-tuned LoRA model is subpar to full model (with a grain of salt since the dataset is very small) and the benefits of using LoRA are unclear.

If we do decide to use Lightning callbacks to support PEFT and, by extension, LoRA, I would suggest hiding the callback interfaces from users. Instead, `FoundationModel` could accept parameters `unfreeze_patterns` and `lora_config` and initialise the corresponding callbacks under the hood. The other `freeze_patterns` can be ignored for simplicity, assuming the "not-unfrozen" weights are always frozen.

It would be intuitive if foundation models like Chronos2 can maintain a simple declaration pattern for fine-tuning like this:

```py
lora_config = LoraConfig(
    ...
)
model_lora = Chronos2Model(
    input_chunk_length=24,
    output_chunk_length=6,
    enable_finetuning=True,
    n_epochs=50,
    lora_config=lora_config,
)
```

Let me know what you think.

**Comment by Kurokabe** [comment]
Hi @daidahao 

Thanks for your input!
The example notebook mostly served as a quick how-to guide to enable fine-tuning and not much about chasing the best  performance. That being said, after increasing the number of epoch to 100 (instead of 50) for LoRA fine-tuning, the results are already much better : 

<img width="547" height="469" alt="Image" src="https://github.com/user-attachments/assets/6d3f998d-cfce-404d-a549-37fd56f4a2cf" />
But still, it needs to be taken with a grain of salt. I did't inspect the training loss or any kind of proper training methodology, so there might be some risks of overfitting.

I like the idea that full and partial fine-tuning could already be integrated within `FoundationModel`. I'll also check with @dennisbader what he thinks of this solution and comeback to you soon :)

**Comment by daidahao** [comment]
> I like the idea that full and partial fine-tuning could already be integrated within FoundationModel. 

@Kurokabe 

Thank you for the updates.

Yes, I believe that partial fine-tuning can be achieved without using `ModelTransformCallback`. If we simply set  `requires_grad=False` for frozen parameters before `PLForecastingModule.train_step()` is called, the gradients won't be computed on those parameters. As a result, those parameters will not be updated. There are a number of hooks we could override for that purpose.

**Comment by daidahao** [comment]
Congratulations! @Kurokabe @dennisbader

**Comment by dennisbader** [comment]
Thanks to you @Kurokabe and @daidahao :) 

Also, release 0.42.0 is coming in a couple of hours 🥳

**Comment by Kurokabe** [comment]
Thank you @daidahao and @dennisbader for the great review and advice 😄
